Tokenize/src/token.jl

```diff
@@ -82,98 +62,31 @@ struct RawToken <: AbstractToken
+function Token(kind::Kind, startbyte::Int, endbyte::Int)
+    Token(kind, startbyte, endbyte, NO_ERR, false, false)
 end
-RawToken() = RawToken(ERROR, 0, 0, UNKNOWN, false, false)
+Token() = Token(ERROR, 0, 0, UNKNOWN, false, false)
 
 
-const _EMPTY_TOKEN = Token()
-const _EMPTY_RAWTOKEN = RawToken()
-EMPTY_TOKEN(::Type{Token}) = _EMPTY_TOKEN
-EMPTY_TOKEN(::Type{RawToken}) = _EMPTY_RAWTOKEN
+const _EMPTY_RAWTOKEN = Token()
+EMPTY_TOKEN(::Type{Token}) = _EMPTY_RAWTOKEN
 
-function kind(t::AbstractToken)
+function kind(t::Token)
     isoperator(t.kind) && return OP
     iskeyword(t.kind) && return KEYWORD
     return t.kind
 end
-exactkind(t::AbstractToken) = t.kind
-startpos(t::AbstractToken) = t.startpos
-endpos(t::AbstractToken) = t.endpos
-startbyte(t::AbstractToken) = t.startbyte
-endbyte(t::AbstractToken) = t.endbyte
-function untokenize(t::Token)
-    if t.kind == IDENTIFIER || isliteral(t.kind) || t.kind == COMMENT || t.kind == WHITESPACE || t.kind == NEWLINE_WS || t.kind == ERROR
-        return t.val
-    elseif iskeyword(t.kind)
-        return lowercase(string(t.kind))
-    elseif isoperator(t.kind)
-        if t.dotop
-            str = string(".", UNICODE_OPS_REVERSE[t.kind]) 
-        else 
-            str = string(UNICODE_OPS_REVERSE[t.kind]) 
-        end 
-        return string(str, t.val) 
-    elseif t.kind == LPAREN
-        return "("
-    elseif t.kind == LSQUARE
-        return "["
-    elseif t.kind == LBRACE
-        return "{"
-    elseif t.kind == RPAREN
-        return ")"
-    elseif t.kind == RSQUARE
-        return "]"
-    elseif t.kind == RBRACE
-        return "}"
-    elseif t.kind == DQUOTE
-        return "\""
-    elseif t.kind == TRIPLE_DQUOTE
-        return "\"\"\""
-    elseif t.kind == BACKTICK
-        return "`"
-    elseif t.kind == TRIPLE_BACKTICK
-        return "```"
-    elseif t.kind == AT_SIGN
-        return "@"
-    elseif t.kind == COMMA
-        return ","
-    elseif t.kind == SEMICOLON
-        return ";"
-    else
-        return ""
-    end
-end
-
-function untokenize(t::RawToken, str::String)
-    String(codeunits(str)[1 .+ (t.startbyte:t.endbyte)])
-end
+exactkind(t::Token) = t.kind
 
-function untokenize(ts)
-    if !(eltype(ts) <: AbstractToken)
-        throw(ArgumentError("element type of iterator has to be Token"))
-    end
-    io = IOBuffer()
-    for tok in ts
-        write(io, untokenize(tok))
-    end
-    return String(take!(io))
-end
+startbyte(t::Token) = t.startbyte
+endbyte(t::Token) = t.endbyte
 
 
-function Base.show(io::IO, t::Token)
-    start_r, start_c = startpos(t)
-    end_r, end_c = endpos(t)
-    str = kind(t) == ENDMARKER ? "" : escape_string(untokenize(t))
-    print(io, rpad(string(start_r, ",", start_c, "-", end_r, ",", end_c), 17, " "))
-    print(io, rpad(kind(t), 15, " "))
-    print(io, "\"", str, "\"")
+function untokenize(t::Token, str::String)
+    String(codeunits(str)[1 .+ (t.startbyte:t.endbyte)])
```

Comment:
This is a nice way to express this operation with `codeunits`.  I'm usually messing with `previndex` for this.